Client/Frontend/Browser/Search/BraveSearchManager.swift

```diff
@@ -0,0 +1,293 @@
+    
+    static func isValidURL(_ url: URL) -> Bool {
+        let validURLs = AppConstants.buildChannel.isPublic ?
+            ["search.brave.com"] : ["search.brave.com", "search-dev.brave.com"]
+        
+        return validURLs.contains(url.host ?? "")
+    }
+    
+    init?(url: URL, cookies: [HTTPCookie]) {
+        if !Self.isValidURL(url) {
+            return nil
+        }
+        
+        // Check if the request has a valid query string to check against.
+        guard let components = URLComponents(url: url, resolvingAgainstBaseURL: false),
+              let queryItem = components.valueForQuery("q") else { return nil }
+        
+        self.url = url
+        self.query = queryItem
+        self.domainCookies = cookies.filter { $0.domain == url.host }
+        if domainCookies.first(where: { $0.name == "fallback" })?.value != "1" {
+            return nil
+        }
+    }
+    
+    /// A call is made to the Brave Search api with query we performed.
+    /// It returns  details that are used in call to the fallback search engine.
+    func shouldUseFallback(completion: @escaping (BackupQuery?) -> Void) {
+        guard var canAnswerURLComponents =
+                URLComponents(string: url.domainURL.absoluteString) else {
+            completion(nil)
+            return
+        }
+        
+        canAnswerURLComponents.scheme = "https"
+        canAnswerURLComponents.path = "/api/can_answer"
+        canAnswerURLComponents.queryItems = [.init(name: "q", value: query)]
+        
+        guard let url = canAnswerURLComponents.url else {
+            completion(nil)
+            return
+        }
+        
+        var request = URLRequest(url: url,
+                                 cachePolicy: .reloadIgnoringLocalAndRemoteCacheData,
+                                 timeoutInterval: 5)
+
+        let cookieStorage = HTTPCookieStorage()
+        domainCookies.forEach { cookieStorage.setCookie($0) }
+        
+        let domainCookies = domainCookies.filter { $0.domain == canAnswerURLComponents.host }
+        let headers = HTTPCookie.requestHeaderFields(with: domainCookies)
+        headers.forEach {
+            request.setValue($0.value, forHTTPHeaderField: $0.key)
+        }
+        
+        // Important, URLSessionDelegate must have been implemented here
+        // to handle request authentication.
+        URLSession(configuration: .ephemeral, delegate: self, delegateQueue: .main)
+            .dataTaskPublisher(for: request)
+            .tryMap { output -> BackupQuery in
+                guard let response = output.response as? HTTPURLResponse,
+                      response.statusCode >= 200 && response.statusCode < 300 else {
+                    throw "Invalid response"
+                }
+     
+                let canAnswerResponse = try JSONDecoder().decode(BackupQuery.self, from: output.data)
+                
+                return canAnswerResponse
```

Comment:
return `output.data` instead of `BackupQuery` here, then add `.decode(type: BackupQuery.self, decoder: JSONDecoder())` to the chain

i.e.

```
.tryMap { output -> Data 
  ...
}
.decode(type: BackupQuery.self, decoder: JSONDecoder())
.sink(...)
```